presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveNullRowInAggregation.java

```diff
@@ -0,0 +1,116 @@
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.relational.FunctionResolution;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.facebook.presto.SystemSessionProperties.isRemoveNullRowInAggregation;
+import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.plan.AggregationNode.Aggregation;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.AND;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.IS_NULL;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+import static java.util.Objects.requireNonNull;
+
+public class RemoveNullRowInAggregation
+        implements Rule<AggregationNode>
+{
+    private static final Pattern<AggregationNode> PATTERN = aggregation();
+
+    private final StandardFunctionResolution functionResolution;
+    private final FunctionAndTypeManager functionManager;
+
+    public RemoveNullRowInAggregation(FunctionAndTypeManager functionAndTypeManager)
+    {
+        requireNonNull(functionAndTypeManager, "functionManager is null");
+        this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+        this.functionManager = functionAndTypeManager;
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isRemoveNullRowInAggregation(session);
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregationNode, Captures captures, Context context)
+    {
+        PlanNode source = aggregationNode.getSource();
+        if (context.getLookup().resolveGroup(source).findFirst().filter(node -> node instanceof FilterNode)
+                .isPresent()) {
+            return Result.empty();
+        }
+        Collection<Aggregation> aggregations = aggregationNode.getAggregations().values();
+
+        //extract all the aggregation arguments
+        Set<RowExpression> rowExpressionSet = new HashSet<>();
+        for (Aggregation aggregation : aggregations) {
+            if (functionManager.getFunctionMetadata(aggregation.getFunctionHandle()).isCalledOnNullInput()) {
+                return Result.empty();
+            }
+            rowExpressionSet.addAll(aggregation.getCall().getArguments());
+        }
+
+        if (rowExpressionSet.isEmpty()) {
+            return Result.empty();
+        }
+
+        //Form all the ISNULL form
+        ImmutableList.Builder<RowExpression> nullFormBuilder = ImmutableList.builder();
```

Comment:
There are some util functions here https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/relational/Expressions.java

Consider reusing them for the RowExpression constructor below

Reply:
fixed

Reply:
If the rule applies to the FilterNode source, then the rule will be applied infinitely. Thanks for your suggestion, I utilized the util functions as you mentioned.